main.c

```diff
@@ -235,17 +236,76 @@ static bool path_exists(const char *path) {
 	return path && access(path, R_OK) != -1;
 }
 
-static const char *get_output_dir(void) {
-	static const char *output_dirs[] = {
-		"GRIM_DEFAULT_DIR",
-		"XDG_PICTURES_DIR",
-	};
-
-	for (size_t i = 0; i < sizeof(output_dirs) / sizeof(char *); ++i) {
-		const char *path = getenv(output_dirs[i]);
-		if (path_exists(path)) {
-			return path;
+static const char *get_xdg_pictures_dir(void) {
+	char *home_dir = getenv("HOME");
+	if (home_dir == NULL) {
+		return NULL;
+	}
+
+	char *config_file;
+	char *config_home = getenv("XDG_CONFIG_HOME");
+	if (config_home == NULL || config_home[0] == 0) {
+		config_file = (char *)malloc(strlen(home_dir) + strlen("/.config/user-dirs.dirs") + 1);
+		if (config_file == NULL) {
+			return NULL;
+		}
+		strcpy(config_file, home_dir);
+		strcat(config_file, "/.config/user-dirs.dirs");
+	} else {
+		config_file = (char *)malloc(strlen(config_home) + strlen("/user-dirs.dirs") + 1);
+		if (config_file == NULL) {
+			return NULL;
+		}
+		strcpy(config_file, config_home);
+		strcat(config_file, "/user-dirs.dirs");
+	}
+
+	FILE *file = fopen(config_file, "r");
+	free(config_file);
+	if (file == NULL) {
+		return NULL;
+	}
+
+	char *line = NULL;
+	size_t line_size = 0;
+	ssize_t nread;
+	char *pictures_dir = NULL;
+	while ((nread = getline(&line, &line_size, file)) != -1) {
+		if (line[nread - 1] == '\n') {
+			line[--nread] = '\0';
 		}
+
+		if (strlen(line) == 0 || line[0] == '#') {
+			continue;
+		}
+
+		size_t i = 0;
+		while (line[i] != '\0' && line[i] != ' ') {
+			i++;
+		}
+		if (strncmp(line, "XDG_PICTURES_DIR=", 17) == 0) {
+			char* line_remaining = line + 17;
+			wordexp_t p;
+			if (wordexp(line_remaining, &p, WRDE_UNDEF) == 0) {
+				pictures_dir = strdup(p.we_wordv[0]);
```

Comment:
FYI, the code has a memory leak where it leaks the old value of `pictures_dir` if the file contains multiple lines starting with with `XDG_PICTURES_DIR=` . (The `xdg-user-dir-lookup` code has the same issue, see https://gitlab.freedesktop.org/xdg/xdg-user-dirs/-/merge_requests/5/diffs .)

(Beyond that, everything looks good.)

Reply:
Also, style: `char* line_remaining` should be `char *line_remaining`